flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java

```diff
@@ -732,9 +745,7 @@ public void run() {
 				throw new RuntimeException(discoveryLoopError);
 			}
 		} else {
-			// won't be using the discoverer
-			partitionDiscoverer.close();
-
+			// partitionDiscoverer is already closed in open method
```

Comment:
Note that it isn't safe to directly close the partition discoverer as it isn't thread-safe (specifically, the Kafka client isn't thread safe). Would need to consider that when cleaning up the partition discoverer in the finally block.

Please see the comments in the `FlinkKafkaConsumerBase.cancel()` method for an example.
If the partition discoverer is concurrently being used by the discovery thread, it must be invoked `wakeup()` instead.

Reply:
It wouldn't change the existing behaviour except for the case where `kafkaFetcher.runFetchLoop` throws an exception where we don't close the `partitionDiscoverer` in line 721. If the `kafkaFetcher` fails, then it shouldn't matter whether the `discoveryLoopThread` also fails or not, right?

Reply:
I am not sure if it is necessary to wrap `runFetchLoop` with try-finally.
1) when discovery disabled, we already take care of it in open method
2) when discovery enabled, cancel() takes care of it. what if some other place in run method throws an exception? we always need cancel() to clean up resource if necessary.

Reply:
Well, I think you've raised a very good point here @stevenzwu. What happens if partition discovery is enabled and the creation of the `kafkaFetcher` fails? Since `close` won't close the partition discoverer, there will be another leak. The same happens if `kafkaFetcher.runFetchLoop` fails in the case of enabled `partitionDiscoverer`.

I think we should rethink the lifecycle management of the `partitionDiscoverer` a bit better. What if we say that the `partitionDiscoverer` will only be closed in the `close` method which is called by the `Task's` main thread. In order to avoid the exceptions we first have to wait on the discoveryLoop's termination. But then it should be safe to close it.